### PR TITLE
Add AnyOf and AllOf matcher

### PR DESCRIPTION
**Pull request by jon-betts**
Matchers which can be used to match any or all of their contents against a target.

This is a merge request of two halves:

 * Added the new matchers and tests in commit 1
 * Moving things about in commit 2

### The movement in commit 2

The problem

 * `AllOf` could easily be added as `Any.of()`
 * `AnyOf` doesn't really have a similar home
 * `Any` can't be easily imported by matches as it imports all of them

The solution:

 * Take the matcher part out of `Any` into `h_matchers.matcher.anything.AnyThing`
 * Move `Any` from `h_matchers.matcher` to `h_matchers.interface`
 * Add a new `All` in `h_matchers.interface`

### CODE REVIEW

File: src/h_matchers/matcher/combination.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+class AnyOf(Matcher):
+    """Match any one of a series of options."""
+
+    def __init__(self, options):
+        options = list(options)  # Coerce generators into concrete list
+
+        super().__init__(
+            f"* any of {options} *",
+            lambda other: any(other == option for option in options),
```
Comment (seanh) [inline src/h_matchers/matcher/combination.py]: How does `any(other == option for option in options)` differ from `other in options`? If there's no difference, then I wonder whether `AnyOf` is really needed, since `<thing> in <container>` is pretty easy to type?

File: src/h_matchers/matcher/combination.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+class AllOf(Matcher):
+    """Match only when all of a series of options match."""
+
+    def __init__(self, options):
+        options = list(options)  # Coerce generators into concrete list
+
+        super().__init__(
+            f"* all of {options} *",
+            lambda other: all(other == option for option in options),
```
Comment (seanh) [inline src/h_matchers/matcher/combination.py]: So `other == All.of(options)` is equivalent to `all(other == option for option in options)`, just slightly easier to type?

File: src/h_matchers/matcher/combination.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            f"* any of {options} *",
+            lambda other: any(other == option for option in options),
+        )
+
+
+class AllOf(Matcher):
+    """Match only when all of a series of options match."""
+
+    def __init__(self, options):
+        options = list(options)  # Coerce generators into concrete list
```
Comment (seanh) [inline src/h_matchers/matcher/combination.py]: Why is this needed?

File: src/h_matchers/matcher/combination.py (hunk 1 above)
Comment (jon-betts) [reply src/h_matchers/matcher/combination.py]: So I think you are totally right about these being equivalent, but forgetting the original motivation for the matchers.

The statements `blah in blah` and the `any()` above are equivalent (and I'll replace that) but neither are the same as an `AnyOf` instance because they don't test as _equal_ to `other`. Consider the following:

`assert [1, 2, 3] == [1, AnyOf([2, 4]), 3]`

I don't see how you'd write that using `in` in the same way. If you are just using `AnyOf` as a roundabout way of calling the function, then you should probably be using `in` directly. The same goes for `callable` and many of the other things we wrap.

The point is `other in options` does not test equal to `other`, which a matcher does, plus:

 * We can have a nice string message
 * We also are `instanceof(Matcher)` which will be quite handy later on

File: src/h_matchers/matcher/combination.py (hunk 2 above)
Comment (jon-betts) [reply src/h_matchers/matcher/combination.py]: See comment above

File: src/h_matchers/matcher/combination.py (hunk 3 above)
Comment (jon-betts) [reply src/h_matchers/matcher/combination.py]: This is because a matcher can be assigned to a variable and used multiple times. If you don't do this, then the first time the matcher is compared the generator is exhausted and all future comparisons fail.
